src/server/pps/server/api_server.go

```diff
@@ -458,32 +524,93 @@ func (a *apiServer) Version(version int64) error {
 	return nil
 }
 
-func (a *apiServer) AddShard(shard uint64, version int64) error {
-	pipelineInfos, err := a.ListPipeline(context.Background(), &ppsclient.ListPipelineRequest{})
+func (a *apiServer) AddShard(shard uint64) error {
+	persistClient, err := a.getPersistClient()
 	if err != nil {
 		return err
 	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	a.shardCancelFuncsLock.Lock()
+	defer a.shardCancelFuncsLock.Unlock()
+	if _, ok := a.shardCancelFuncs[shard]; ok {
+		return fmt.Errorf("shard %d is being added twice; this is likely a bug", shard)
+	}
+	a.shardCancelFuncs[shard] = cancel
+
+	client, err := persistClient.SubscribeNewPipeline(ctx, &persist.SubscribeNewPipelineRequest{})
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		pipelineChan := make(chan *ppsclient.PipelineInfo)
+		go func() {
+			for {
+				pipelineInfo, err := client.Recv()
+				if err != nil {
+					protolion.Printf("error receiving new pipeline: %v", err)
+				}
+				pipelineChan <- newPipelineInfo(pipelineInfo)
```

Comment:
So it seems like if there's an error from `client.Recv()` then we're going to have `pipelineInfo == nil` which means that `newPipelineInfo(pipelineInfo)` will crash.

This function can be made a lot simpler. You can just call `client.Recv()` in a for loop. If `cancel` has been called then `client.Recv()` will error and you can just break the loop there.